### PR TITLE
Dynamic clk switch

### PR DESCRIPTION
**Pull request by keyonjie**
cavs: discard CAVS_LPRO and set CAVS_USE_LPRO_IN_WAITI by default
    
    With dynamic clock switching applied, we don't need to configure
    CAVS_LPRO explicitly, instead, we will use HPRO by default, and switch
    to use LPRO in low power scenarios (e.g. WFI, D0I3 with KWD). Switching
    to use LPRO in waiti or not is configured by CAVS_USE_LPRO_IN_WAITI
    item, and use LPRO in waiti by default for power saving.
    
    The commit also refine the clock switching logic as below:
    1. use the active_freq_idx to save the active (non-WFI) clock index.
    2. refine helper platform_set_active_clock() to switch the active clock
    and update the active_freq_idx together.
    3. for WFI related clock switching, don't update the active_freq_idx,
    but just do the real switching and switching back.
    4. implement platform_clock_waiti_entry() and
    platform_clock_waiti_exit() for the clock switching, and update the
    active clock before each WFI entry.
    
    This is verified works on TGL platforms, we also need testing on JSL
    with it.

### CODE REVIEW

File: src/platform/intel/cavs/platform.c [hunk 1]
```diff
@@ -146,13 +146,8 @@ EXT_MAN_PORT(
...
 
 #if CONFIG_CANNONLAKE || CONFIG_ICELAKE || CONFIG_TIGERLAKE
-#if CONFIG_CAVS_LPRO
-#define CAVS_DEFAULT_RO		SHIM_CLKCTL_RLROSCC
-#define CAVS_DEFAULT_RO_FOR_MEM	SHIM_CLKCTL_OCS_LP_RING
-#else
-#define CAVS_DEFAULT_RO		SHIM_CLKCTL_RHROSCC
-#define CAVS_DEFAULT_RO_FOR_MEM	SHIM_CLKCTL_OCS_HP_RING
-#endif
+#define CAVS_ROS	(SHIM_CLKCTL_RHROSCC | SHIM_CLKCTL_RLROSCC)
```
Comment (mmaka1) [inline src/platform/intel/cavs/platform.c:149]: Why is CONFIG_CAVS_LPRO support removed here? 
You could request LPRO in #if and request both in #else to get the same result with CAVS_LPRO removed from the TGL config and still support LPRO-only configuration.

File: src/platform/intel/cavs/platform.c [hunk 2]
```diff
@@ -432,15 +427,15 @@ int platform_init(struct sof *sof)
...
+	while (!((shim_read(SHIM_CLKSTS) & CAVS_ROS) == CAVS_ROS))
 		idelay(16);
 
 	shim_write(SHIM_CLKCTL,
-		   CAVS_DEFAULT_RO | /* Request configured RING Osc */
-		   CAVS_DEFAULT_RO_FOR_MEM | /* Select configured
-					     * RING Oscillator Clk for memory
-					     */
+		   CAVS_ROS | /* Request configured RING Osc */
+		   CAVS_ROS_FOR_MEM | /* Select configured
```
Comment (mmaka1) [inline src/platform/intel/cavs/platform.c:436]: Could you explain why single RO selection is changed to selecting both ROs for memory?

File: src/arch/xtensa/configs/override/jasperlake_chrome.config [hunk 3]
```diff
@@ -1,2 +1,2 @@
 CONFIG_CAVS_LPS=y
-CONFIG_CAVS_LPRO=y
+CONFIG_CAVS_USE_LPRO_IN_WAITI=y
```
Comment (mmaka1) [inline src/arch/xtensa/configs/override/jasperlake_chrome.config:2]: Why are CNL and JSL changed on TGL specific branch? Such changes should be handled by separate commits.

File: src/platform/Kconfig [hunk 4]
```diff
@@ -296,17 +296,9 @@ config CAVS_LPS
 	  This option is required to support S0ix/D0ix mode
 	  on cAVS platforms.
 
-config CAVS_LPRO
-	bool "Enable low power ring oscillator as the default clock"
-	default n
-	depends on CAVS && !CAVS_VERSION_1_5
-	help
-	  Select if you want to use 120MHz LPRO as the DSP clock source.
```
Comment (mmaka1) [inline src/platform/Kconfig:304]: The description is no longer valid and should be changed.
But are not we able to support down scaled LPRO-only configuration? Could you explain why it is completely removed?

File: src/platform/intel/cavs/lib/clk.c [hunk 5]
```diff
@@ -124,8 +136,31 @@ void platform_clock_on_waiti(void)
 	}
 }
 
+void platform_clock_waiti_exit(void)
```
Comment (mmaka1) [inline src/platform/intel/cavs/lib/clk.c:139]: Pls explain why is this piece of code duplicated from platform_clock_on_wakeup() (which is called on waiti exit) and put inside the idle task?

File: src/platform/intel/cavs/lib/clk.c [hunk 6]
```diff
@@ -102,15 +102,27 @@ void platform_clock_on_wakeup(void)
 }
 #endif
 
-void platform_clock_on_waiti(void)
+void platform_clock_waiti_entry(void)
```
Comment (mmaka1) [inline src/platform/intel/cavs/lib/clk.c:105]: Note that the old function is being fixed by #3113 on master branch. Is there any rationale behind changing the name while this new piece of deferred clock setup is added?

File: src/platform/intel/cavs/lib/clk.c [hunk 7]
```diff
@@ -101,24 +100,69 @@ void platform_clock_on_wakeup(void)
...
 {
-	int freq_idx = get_cpu_current_freq_idx();
+	int freq_idx;
+	int target_idx;
+
+	/* update the active clock according to the pm runtime state */
+	if (pm_runtime_is_active(PM_RUNTIME_DSP, PLATFORM_MASTER_CORE_ID) &&
+	    *cache_to_uncache(&active_freq_idx) != CPU_HPRO_FREQ_IDX)
+		platform_set_active_clock(CPU_HPRO_FREQ_IDX);
+	else if (*cache_to_uncache(&active_freq_idx) != CPU_LPRO_FREQ_IDX)
```
Comment (slawblauciak) [inline src/platform/intel/cavs/lib/clk.c:114]: Shouldn't you check for !pm_runtime_is_active here? If  pm_runtime_is_active returns true and active clock is already set to HPRO, I believe we'd land in that else if block and put the DSP back in LPRO wrongly

**Comment by keyonjie** [comment]
On 7/8/20 8:42 PM, Sławomir Błauciak wrote:
> @slawblauciak commented on this pull request.
> 
> 
> 
>>   {
> -	int freq_idx = get_cpu_current_freq_idx();
> +	int freq_idx;
> +	int target_idx;
> +
> +	/* update the active clock according to the pm runtime state */
> +	if (pm_runtime_is_active(PM_RUNTIME_DSP, PLATFORM_MASTER_CORE_ID) &&
> +	    *cache_to_uncache(&active_freq_idx) != CPU_HPRO_FREQ_IDX)
> +		platform_set_active_clock(CPU_HPRO_FREQ_IDX);
> +	else if (*cache_to_uncache(&active_freq_idx) != CPU_LPRO_FREQ_IDX)
> 
> Shouldn't you check for !pm_runtime_is_active here? If  pm_runtime_is_active returns true and active clock is already set to HPRO, I believe we'd land in that else if block and put the DSP back in LPRO wrongly

Correct, thanks for spotting this, will update soon.

>

File: src/arch/xtensa/configs/override/jasperlake_chrome.config (hunk 3 above)
Comment (keyonjie) [reply src/arch/xtensa/configs/override/jasperlake_chrome.config:2]: I thought even it is TGL specific branch, we should at least make CNL/JSL passed building, no?

File: src/ipc/handler.c [hunk 8]
```diff
@@ -677,6 +677,13 @@ static int ipc_pm_gate(uint32_t header)
 	else
 		pm_runtime_enable(PM_RUNTIME_DSP, PLATFORM_MASTER_CORE_ID);
 
+#if !CONFIG_CAVS_LPRO
+	if (pm_runtime_is_active(PM_RUNTIME_DSP, PLATFORM_MASTER_CORE_ID))
```
Comment (mmaka1) [inline src/ipc/handler.c]: If there is a risk of having unbalanced pm_gate IPC requests, pm_runtime_disable/enable calls should be protected with client bitmap, not a simple ref-counter (see how the put/get on the DSP resource is protected).

File: src/platform/intel/cavs/platform.c (hunk 1 above)
Comment (keyonjie) [reply src/platform/intel/cavs/platform.c:149]: I can't imagine the usage of the scale down scenario, to avoid confusion, I just removed it.

File: src/platform/intel/cavs/platform.c (hunk 2 above)
Comment (keyonjie) [reply src/platform/intel/cavs/platform.c:436]: It was only a temp to combine the configuration of the DEFAULT RO for both CAVS_LPRO selected and deselected.

File: src/platform/intel/cavs/lib/clk.c (hunk 6 above)
Comment (keyonjie) [reply src/platform/intel/cavs/lib/clk.c:105]: I didn't notice what #3113 is changing, the reason I am changing the naming is just to make the _xx_entry() and _xx_exit() symmetric and are invoked in the platform_wait_for_interrupt() together.

File: src/platform/intel/cavs/lib/clk.c (hunk 5 above)
Comment (keyonjie) [reply src/platform/intel/cavs/lib/clk.c:140]: I didn't know how the platform_clock_on_wakeup() is used, so I didn't want to touch or change it. But yes, if they(both entry() and exit()) can be put inside the waiti, it will be better.

File: src/platform/Kconfig (hunk 4 above)
Comment (lbetlej) [reply src/platform/Kconfig:304]: @keyonjie  I'll second @mmaka1 that (at least for a while) the ability to select LPRO-only should remain as a debug option. Default configuration shall be dynamic clock switching. I'm OK with dynamic clock switching as the only option on https://github.com/thesofproject/sof/tree/tgl-007-drop-stable and a default one but I'd expect that up-streamed code to master will preserve ability to select LPRO-only support.

File: src/platform/intel/cavs/platform.c (hunk 1 above)
Comment (ranj063) [reply src/platform/intel/cavs/platform.c:149]: @keyonjie it looks like we really only have 2 options right? #if CONFIG_CAVS_LPRO is enabled, we do not enable HPRO but LPRO is enabled whether this config is selected or not. So, we should keep the #if CONFIG_CAVS_LPRO support here

File: src/ipc/handler.c [hunk 9]
```diff
@@ -677,6 +677,13 @@ static int ipc_pm_gate(uint32_t header)
...
+		platform_set_active_clock(CPU_HPRO_FREQ_IDX);
+	else
+		platform_set_active_clock(CPU_LPRO_FREQ_IDX);
+#endif
```
Comment (ranj063) [inline src/ipc/handler.c]: @mmaka1 @slawblauciak we also send the PM_GATE IPC when the DSP enters D0i3 in S0. The only way  for the FW to know if it is sent during S0 or S0ix is by checking whether the DMA_TRACE is requested to be turned off or not. Do you think it makes a difference whether the switching to LPRO is happening during S0 as well?

File: src/platform/Kconfig (hunk 4 above)
Comment (keyonjie) [reply src/platform/Kconfig:304]: OK, let me preserve it and rename it to CAVS_LPRO_ONLY.

Comment (keyonjie) [reply src/platform/Kconfig:304]: @lbetlej @mmaka1 the refining is here: https://github.com/thesofproject/sof/pull/3205

**Comment by keyonjie** [comment]
tgl-007 release was done, so close this.
